change tuple indexing

brymer-meneses opened the pull request:
this changes the tuple index from `tuple[0]` to `tuple.0` in accordance with the accepted propsal https://github.com/carbon-language/carbon-lang/pull/3646

brymer-meneses committed:
docs: add a note on setting up `clangd`

brymer-meneses committed:
docs: add a note on setting up `clangd`

zygoloid committed:
Compute specific constant values. (#4128)

When forming a specific (previously called a generic instance), evaluate
the eval block of the generic to determine the values of any constants
used in that specific. The majority of the work here is updating
eval.cpp so that it can use the results of prior evaluations in the same
block when computing later values.

Include the computed results in the formatted SemIR output.

---------

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

zygoloid committed:
Build an evaluation block for the definition region of a generic. (#4131)

zygoloid committed:
Rebuild the type of constants during evaluation. (#4138)

When evaluating in a generic context, a constant with a symbolic type
might evaluate to a constant with a concrete type (or a more specific
symbolic type). This can't actually happen yet given the current state
of the toolchain, as far as I can determine, so this is more just a
refactoring for now, but will be relied upon by future generics work.

---------

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>
Co-authored-by: Carbon Infra Bot <carbon-external-infra@google.com>

jonmeow committed:
Add basic output of where memory is stored after a compile. (#4136)

The output is really basic, I'm just adding this to help track how
memory is allocated.

```
---
filename:        'check/testdata/expr_category/in_place_tuple_init.carbon'
source_:
  used_bytes:      8057
  reserved_bytes:  8057
tokens_.allocator_:
  used_bytes:      0
  reserved_bytes:  0
tokens_.token_infos_:
  used_bytes:      1040
  reserved_bytes:  2032

(eliding)

value_stores_.string_literals_.set_:
  used_bytes:      320
  reserved_bytes:  320
Total:
  used_bytes:      20609
  reserved_bytes:  29437
...
```

---------

Co-authored-by: josh11b <15258583+josh11b@users.noreply.github.com>

gysddn committed:
Do not query interface's self_param_id unless defined when importing (#4137)

Querying a local constant with an invalid instruction triggers a crash,
this will prevent self_param_id to be used unless it is defined.

Closes #4071 and #4080

This will only fix the crash. The scenario where a declaration-only
interface is imported then defined still gives an error message, but it
won't crash this time.

zygoloid committed:
Substitute into generic class and interface definitions when we require them to be fully defined. (#4139)

When referring to a constant within a specific, such as a field of a
generic class, use that specific version of the constant's value.

chandlerc committed:
Enable the more effective version of `-Wnon-virtual-dtor`. (#4142)

Most of this is enabled by default, but there is some that needs an
explicit flag.

Avoiding `-Wnon-virtual-dtor` itself for now because that warning can
require changes that carry overhead such as having an extra, unused
destructor entry in the vtable. Hopefully we don't have too many folks
who need our code to be `-Wnon-virtual-dtor` clean.

chandlerc committed:
Directly use TCMalloc rather than the system malloc on Linux. (#4133)

This improve the toolchain's performance by about 10%.

It will also allow us to leverage TCMalloc's extensions to do heap
profiling and get other information about how efficiently we're using
the heap.

Note that currently this causes all of our builds to produce a warning
due to an issue with `rules_python` and multiple modules registering
python toolchains:
https://github.com/bazelbuild/rules_python/issues/1818

This is also only enabled on Linux as there is no support for other OSes
at the moment.

zygoloid committed:
Mark some leaf classes `final` to suppress `-Wnon-virtual-dtor`. (#4141)

jonmeow committed:
Use verbose formatting of instructions on crash messages. (#4125)

Changes crash messages to start printing verbose forms of instructions,
rather than just the ID. Fixes some indentation issues with stacks. Also
switches unexpected inst formatting, because now there are lots, and
it'd be helpful to know where they are.

This uses a pimpl pattern for Formatter due to the number of member
functions on Formatter. Maybe we should refactor that, but this didn't
feel like a good place to do so.

Note, I have two concerns about this change... to note them here, to
make sure others are considering them when evaluating the
implementation:

1. Some instructions are very verbose to print, as evidenced by the
fn_decl printing (which includes function params) or scope printing
(which includes scope members).
- I'm not sure whether there's a way to simply reduce this, as it seems
essential to the requested printing of instructions.
- Long-term, we may at least want to limit the number of lines printed
here. However, I've already spent a fair amount of time here and I think
it's in a good state to evaluate.
2. Increased complexity in the crash handler may result in crash
messages failing to generate.
- For example, a crash in Formatter (and its deps, such as InstNamer or
location handling) prevents a stack from being printed. I'm pretty sure
I've written crashes in Formatter before.

Here's an example crash snippet (generated by adding a crash inside
`return` handling) before:

```
2.	NodeStack:
	0.	FunctionDefinitionStart -> function2
	1.	ReturnStatementStart -> no value
	2.	IntLiteral -> inst+26
inst_block_stack_:
	0.	block<invalid>	{inst+0, inst+1, inst+2, inst+23}
	1.	block9	{inst+26}
param_and_arg_refs_stack:
args_type_info_stack_:
```

And after:

```
2.	Check::Context
          NodeStack:
            0. FunctionDefinitionStart: function2
            1. ReturnStatementStart: no value
            2. IntLiteral:
              unexpected.inst+26.loc12_10: i32 = int_literal 0 [template = constants.%.2]
          inst_block_stack_:
            0. block<invalid> {
                package: <namespace> = namespace [template] {
                  .Core = unexpected.inst+2
                  .F = unexpected.inst+23.loc11_22
                }
                unexpected.inst+1 = import Core
                unexpected.inst+2: <namespace> = namespace unexpected.inst+1, [template] {}
                unexpected.inst+23.loc11_22: %F.type = fn_decl @F [template = constants.%F] {
                  unexpected.inst+9.loc11_9: init type = call constants.%Bool() [template = bool]
                  unexpected.inst+10.loc11_9: type = value_of_initializer unexpected.inst+9.loc11_9 [template = bool]
                  unexpected.inst+11.loc11_9: type = converted unexpected.inst+9.loc11_9, unexpected.inst+10.loc11_9 [template = bool]
                  unexpected.inst+12.loc11_6: bool = param b
                  @F.%b: bool = bind_name b, unexpected.inst+12.loc11_6
                  unexpected.inst+19.loc11_18: init type = call constants.%Int32() [template = i32]
                  unexpected.inst+20.loc11_18: type = value_of_initializer unexpected.inst+19.loc11_18 [template = i32]
                  unexpected.inst+21.loc11_18: type = converted unexpected.inst+19.loc11_18, unexpected.inst+20.loc11_18 [template = i32]
                  @F.%return: ref i32 = var <return slot>
                }
              }
            1. block9 {
                unexpected.inst+26.loc12_10: i32 = int_literal 0 [template = constants.%.2]
              }
          param_and_arg_refs_stack:
          args_type_info_stack_:
```

zygoloid committed:
Change TypeId to be a thin wrapper around ConstantId. (#4140)

This better follows the principle that types are simply constants of
type `type`, and allows more uniform treatment of types as just another
kind of constant from generics handling.

Use a hash table to map from `TypeId` to information about the complete
type. This makes basic operations on types a bit simpler, and operations
that actually need to access the complete class information a bit more
complex.

zygoloid committed:
Move `GetTypeInInstance` from `Check` to `SemIR`. (#4144)

In preparation for this function being used by other parts of `SemIR`
and by lowering.

zygoloid committed:
Basic testing for generic methods using `Self`. (#4143)

chandlerc committed:
Reserve memory for the identifiers hashtable.  (#4107)

This uses a heuristic reserve to greatly reduce hashtable growth of the
identifiers hashtable. The design of the hashtable itself is optimized
around compact memory use and is especially slow to grow and so this has
an outsized impact.

The heuristic was computed using `scripts/source_stats.py` and looking
at C++ codebases. We may want to periodically re-evaluate it as Carbon
code emerges and we have better data on its distributions of tokens.

This also required fixing the `Reserve` method on `CanonicalValueStore`
that wasn't actually used anywhere and so didn't even compile correctly.
I added it to the relevant unit test so it is at least compiled locally
to its definition.

jonmeow committed:
Abbreviate instruction as inst in formatter. (#4146)

This is just consistency with other abbreviating we're doing.

jonmeow committed:
Try some crash recovery in autoupdate threads. (#4147)

At present, I think if we crash from multiple threads in parallel, it
can lead to the stack trace not being printed out. Using
CrashRecoveryContext here seems to more successfully print a stack trace
on errors, which I'm hoping will ease debugging.

i.e., before:

```
-----------------------------------------------------------------------------
.Please report issues to https://github.com/carbon-language/carbon-lang/issues and include the crash backtrace.
Stack dump:
0.	performing autoupdate for toolchain/check/testdata/alias/no_prelude/import_order.carbon
1.	Program arguments: compile --phase=check --dump-sem-ir --no-prelude-import --exclude-dump-file-prefix=/usr/local/google/home/jperkins/.cache/bazel/_bazel_jper.kins/85deb7d9d96f7e0e80b42618a55969d7/execroot/_main/bazel-out/k8-fastbuild/b.in/toolchain/install/prefix_root/lib/carbon/../../lib/carbon/core a.carbon b.carbon
.CHECK failure at ./toolchain/sem_ir/ids.h:140: is_valid()
CHECK failure at ./toolchain/sem_ir/ids.h:140: is_valid()
external/bazel_tools/tools/test/test-setup.sh: line 328: 1151859 Aborted                 "${TEST_PATH}" "$@" 2>&1
```

(EOF)

after:

```
-----------------------------------------------------------------------------
Please report issues to https://github.com/carbon-language/carbon-lang/issues and include the crash backtrace.
Stack dump:
0.	performing autoupdate for toolchain/check/testdata/alias/no_prelude/import_order.carbon
1.	Program arguments: compile --phase=check --dump-sem-ir --no-prelude-import --exclude-dump-file-prefix=/usr/local/google/home/jperkins/.cache/bazel/_bazel_jperkins/85deb7d9d96f7e0e80b42618a55969d7/execroot/_main/bazel-out/k8-fastbuild/bin/toolchain/install/prefix_root/lib/carbon/../../lib/carbon/core a.carbon b.carbon
..CHECK failure at ./toolchain/sem_ir/ids.h:140: is_valid()
CHECK failure at ./toolchain/sem_ir/ids.h:140: is_valid()
......CHECK failure at ./toolchain/sem_ir/ids.h:140: is_valid()
CHECK failure at ./toolchain/sem_ir/ids.h:140: is_valid()
.....................................................................................CHECK failure at ./toolchain/sem_ir/ids.h:140: is_valid()
.CHECK failure at ./toolchain/sem_ir/ids.h:140: is_valid()
............................ #0 0x000056045ee22d7d llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/local/google/home/jperkins/.cache/bazel/_bazel_jperkins/85deb7d9d96f7e0e80b42618a55969d7/execroot/_main/bazel-out/k8-fastbuild/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x731dd7d)
```

(elided the full stack trace)

jonmeow committed:
Remove irregular digit placement check from NumericLiteral. (#4150)

This changed in #1983 and although we updated explorer, we missed
toolchain.

geoffromer committed:
Rename BindNameInfo to EntityName (#4090)

zygoloid committed:
Remove declaration of function deleted in #4150. (#4151)

jonmeow committed:
Remove already-done no_prelude todo. (#4148)

jonmeow committed:
Use the package/library name in ImportIRId formatting. (#4154)

Also adds import_ir_scope to namespace formatting. I'd done this as an
aid for #4153, and am splitting it out.

---------

Co-authored-by: Geoff Romer <gromer@google.com>

jonmeow committed:
Implement syntactic merge checks for parameters. (#4149)

Note this isn't implementing checking through imports. The parse node
there is harder to access through the context, so would require
examining the entity in order to get the import declaration, to get at
the ImportIR. We also don't have a parse tree attached in that case, and
would need to add one to SemIR::File. But I believe we do want to add
that, so it's explicitly a TODO.

Note GetTokenText re-lexes literal values, so there's a bit of potential
overhead there. Not sure if we want a more efficient manner for
comparing in cases like this.

zygoloid committed:
Rebuild all constants in the eval block. (#4155)

Instead of reusing instructions from the generic entity in the eval
block, rebuild constants in the same way we rebuild types. The previous
attempt to not rebuild these constants assumed that every constant used
in a generic would be built in that generic, and not referenced directly
or referenced from some enclosing scope, which isn't true in practice
and is a fragile assumption in any case.

We could add back some reuse of instructions from the generic -- if we
happen to see the right instruction to build a constant, we could
opportunistically reuse it -- but given the complexity added by doing
so, I'm not pursuing that here.

Now that the eval block for a generic consists of instructions uniquely
owned by that generic, rather than often being shared with another
entity, include the generic in the formatted SemIR output. I'm using the
same scope name for the generic object itself as for the parameterized
class / function / interface, because there are very frequently
references between them and this keeps the IR simpler and more readable,
and avoids needing to invent a second name for the scope.

zygoloid committed:
Remove overeager CHECK. (#4159)

When evaluating within the context of a specific, we can encounter uses
of bindings that are nested within that specific, for example parts of
the declaration of a nested generic. Those bindings should evaluate to
the canonical form of themselves, as they would when evaluating outside
the context of the specific.

Fixes #4157.

brymer-meneses committed:
apply suggestions

brymer-meneses committed:
forgot a semicolon

zygoloid committed:
Convert `EvalContext` into a class. (#4160)

chandlerc committed:
Add a mention of our nightly builds to our `README.md`. (#4158)

This provides an example of downloading and using the nightly buildings
of the toolchain. I've tried to provide appropriate caveats about the
fact that this is very early and not something that's really reliable.
But I wanted folks to know how they can play with the nightly releases
if they're interested.

I've also focused the source build instructions on the toolchain where
we'd be interested in contributions, and the first paragraph on trying
out Carbon in the browser with CE.

---------

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

jonmeow committed:
Improve namespace handling in imports. (#4153)

This implements a few closely related features:

- Starts merging namespaces discovered inside imports.
- Stores results of cross-package name lookup as an entry inside the
scope.
  - Note this is particularly visible with `i32`.
- Moves more of the imported instructions to the import scope.

Note this is primarily for executing the namespace TODO in check.cpp,
which is removed here.
`testdata/namespace/merging_with_indirections.carbon` tests key
behavior.

---------

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

zygoloid committed:
Support for calling non-generic methods in a specific class. (#4156)

Use the specific parameter types for checking, and the specific return
type as the type of the call.

zygoloid committed:
When reentering an interface scope, reintroduce the `Self` parameter. (#4162)

This is not easy to test right now, because it should only really be
visible through `default fn` declarations, which we don't support
properly yet.

jonmeow committed:
Move common entity fields to a 'base' struct. (#4161)

I'd considered moving DeclParams uses over, but when handling qualified
names, there's a parse node instead of an instruction. I did try to
unify a couple other uses though, including adding MergeDefinition. I
expect `interface` will use a little more once it's more completely
implemented, but maybe I'm wrong about that.

zygoloid committed:
Rename "generic instance" to "specific" throughout the toolchain. (#4165)

As discussed in toolchain meeting, we want to avoid overloading the
meaning of "instance", and "specific" was the best name we found. It's a
little unorthodox and inventive, but hopefully over time will become as
unsurprising as the term "generic" is.

brymer-meneses committed:
fix: `clangd: -32001: invalid AST` (#4164)

Ignore the `-march=*` in `compile_commands.json` which causes issues in
Neovim on Apple M2. This fix is from this
[comment](https://github.com/clangd/clangd/issues/1582#issuecomment-1500031372)

zygoloid committed:
Remove caching of return slot usage. (#4163)

The caching isn't buying us much, and is adding complexity and
divergence between the codepaths for generic and non-generic functions.

This means we no longer suppress diagnostics for the second or
subsequent time we call a function with an incomplete return type. If we
want to add that back, it might be worth considering moving the
suppression to `TryToCompleteType` and only diagnosing that a type is
incomplete once, regardless of why we're requiring it to be complete.

jonmeow committed:
Improve notes on MODULE.bazel.lock changes (#4167)

I admit I'm tempted to make a MODULE.bazel.md for these comments, so
that modifying them doesn't trigger a lockfile update, but I don't
really want it at the top level.

jonmeow committed:
Only produce a name scope for a namespace when not merged (#4168)

Addresses the comment on
https://github.com/carbon-language/carbon-lang/pull/4153#discussion_r1690292168
(although maybe with somewhat quirky results, since contents get printed
each time)

zygoloid committed:
Move definition of member of Function to function.cpp where it belongs. (#4170)

zygoloid committed:
Refactor `ReturnTypeInfo` and `InitRepr`. (#4169)

Rename `ReturnInfo` to `ReturnTypeInfo`. Move it and `InitRepr` into
`type_info.h` alongside `ValueRepr`. Replace `ReturnSlot` with
`InitRepr`, and extend `InitRepr` to be able to represent the
incomplete-type case instead of CHECK-failing. Remove `has_return_slot`
from `InitRepr` and instead only provide that as part of
`ReturnTypeInfo`.

jonmeow committed:
Add initial parsing for 'extern library' (#4173)

jonmeow committed:
Refactor some commonality in formatter. (#4171)

Also tries to add comments for things. Note, I'm trying to improve
understandability here, so if you don't think this is helping I can undo
things.

jonmeow committed:
Refactor some check-phase postorder iterator use. (#4175)

Allow directly constructing a PostorderIterator, to get rid of
`tree.postorder(node_id).end()` indirect construction. For ranges that
don't need tree data, make it clearer that they're not validated.

Note, this subtly gets rid of a subtree size use in the
`tree.postorder(node_id).end()` case (to get the discarded `begin()`
value).

jonmeow committed:
Temporarily disable tcmalloc due to compiler-explorer crash (#4177)

Per @axsaucedo on #4176, tcmalloc expects cpu information that
compiler-explorer is lacking in its sandboxing. There's probably a
better fix to be had, this is intended to be temporary.

zygoloid committed:
Fix incorrect `value_id` and location in imported `BindSymbolicName`. (#4178)

Instead of updating the `value_id` on the canonical constant
`BindSymbolicName` to refer to some particular instance of that
constant, create a new instruction, and attach the proper location to
it.

jonmeow committed:
Separate subtree size information from parse nodes. (#4174)

Move subtree sizes over to TreeAndSubtrees, using the different
structure to represent the additional parse work that occurs, as well as
making it clear which functions require the extra information. My intent
is to make it hard to use this by accident.

The subtree size is still tracked during Parse::Tree construction. I
think a lot of that can be cleaned up, although we use it during
placeholder assignment so it may take some work. I wanted to see what
people thought about this before taking action on such a change.

I'm using a 1m line source file generated by #4124 for testing. Command
is `time bazel-bin/toolchain/install/prefix_root/bin/carbon compile
--phase=check --dump-mem-usage ~/tmp/data.carbon`

At head, what I'm seeing is:

```
...
parse_tree_.node_impls_:
  used_bytes:      61516116
  reserved_bytes:  61516116
...
Total:
  used_bytes:      447814230
  reserved_bytes:  551663894
...
1.43s user 0.14s system 99% cpu 1.565 total
```

With `Tree::Verify` disabled completely, it looks like:
```
parse_tree_.node_impls_:
  used_bytes:      41010744
  reserved_bytes:  41010744
...
Total:
  used_bytes:      427308858
  reserved_bytes:  531158522
...
1.20s user 0.13s system 99% cpu 1.332 total
```

Re-enabling just the basic verification (what is now `Tree::Verify`),
I'm seeing maybe 0.05s slower, but that's within noise for my system. I
do see variability in my timing results, and overall I think this is a
0.2s +/- 0.1s improvement versus the earlier (always testing `Extract`
code) implementation. That's opt; debug builds will be unaffected,
because the same checking occurs as before.

Note, the subtree size is a third of the node representation, which is
why I'm showing the decrease in memory usage here.

zygoloid committed:
Import support for generics and specifics (#4179)

Import generics and specifics when they are referenced by imported
entities.

When importing a generic, we import the symbolic constants required by
its eval block, and then rebuild the eval block itself given the list of
constants it needs to compute. This is likely a bit less efficient than
directly importing the contents of the eval block, but avoids needing to
either extend the importer code to be able to import the instructions
that can appear in the eval block or extend the evaluator to cope with
instructions from a different `SemIR::File`.

Importing a symbolic constant is unaffected, and does not yet preserve
the associated generic and index within that generic, so uses of a
generic from an imported IR still don't pick up values from the
specific, but the improved functionality can be seen in the changes to
the SemIR in the testcases.

zygoloid committed:
When importing symbolic constants and types, also import the associated generic and index. (#4180)

A symbolic constant has an instruction to compute the constant value, as
well as potentially also having a generic ID and an index within that
generic to indicate where corresponding values can be found in a
specific. Import those pieces of information when importing such a
constant.

We try to import the generic before we start the main work of importing
the constant, and retry the import process if importing the generic adds
work to the worklist. This means that the first time we import anything
within a generic, we can now perform three passes calling
`TryResolveInst` instead of two, but the first pass is very lightweight
and only looks up and adds a single instruction, so the added overhead
of the extra pass should be minimal.

To avoid introducing cycles when importing a generic function, make the
import of a function declaration build the new `Function`,
`FunctionDecl`, and `FunctionType` in the first pass, like classes and
interfaces do.

dwblaikie committed:
Remove supurfluous/confusing {} around a temporary (#4183)

This was failing to build for me locally with some arbitrary Clang HEAD
host compiler:
```
migrate_cpp/rewriter.cpp:225:3: error: call to member function 'SetReplacement' is ambiguous
  225 |   SetReplacement(expr, {OutputSegment(std::move(text))});
      |   ^~~~~~~~~~~~~~
./migrate_cpp/rewriter.h:141:8: note: candidate function [with T = clang::IntegerLiteral]
  141 |   auto SetReplacement(const T* node, std::vector<OutputSegment> output_segments)
      |        ^
./migrate_cpp/rewriter.h:150:8: note: candidate function [with T = clang::IntegerLiteral]
  150 |   auto SetReplacement(const T* node, OutputSegment segment) -> void {
      |        ^
```
No idea if that's a bug in clang HEAD, but it seemed like removing the
{} simplified the code anyway - so here's that.

zygoloid committed:
Initial rough support for deducing generic arguments in a call to a generic function. (#4184)

brymer-meneses committed:
change tuple indexing

this changes the tuple index from `tuple[0]` to `tuple.0` in accordance
with the accepted propsal https://github.com/carbon-language/carbon-lang/pull/3646

brymer-meneses committed:
remove unused header

brymer-meneses committed:
update forgotten tests

josh11b commented on toolchain/check/testdata/index/fail_tuple_large_index.carbon:
```diff
@@ -10,15 +10,15 @@
 // CHECK:STDERR: var c: i32 = b[1];
 // CHECK:STDERR:              ^~~~
 // CHECK:STDERR:
 var c: i32 = b[1];
```
I'm surprised this wasn't changed to `b.1` like other instances

josh11b commented on toolchain/check/handle_name.cpp:
```diff
@@ -22,6 +23,15 @@ auto HandleParseNode(Context& context, Parse::MemberAccessExprId node_id)
+        PerformTupleIndex(context, node_id, tuple_inst_id, index_inst_id);
+
+    context.node_stack().Push(node_id, tuple_value_inst_id);
+
```
For consistency with this code and other Carbon code.
```suggestion
```

josh11b commented on toolchain/check/handle_name.cpp:
```diff
@@ -52,6 +62,16 @@ auto HandleParseNode(Context& context, Parse::PointerMemberAccessExprId node_id)
+        PerformTupleIndex(context, node_id, tuple_inst_id, index_inst_id);
+
+    context.node_stack().Push(node_id, tuple_value_inst_id);
+
```
```suggestion
```

josh11b commented on toolchain/check/member_access.cpp:
```diff
@@ -391,7 +391,6 @@ auto PerformMemberAccess(Context& context, Parse::NodeId node_id,
 
   return member_id;
 }
```
Should not delete the newline that was there.

```suggestion
}

```

josh11b commented on toolchain/check/testdata/index/index_not_literal.carbon:
```diff
@@ -9,7 +9,10 @@
+// CHECK:STDERR: index_not_literal.carbon:[[@LINE+3]]:14: ERROR: Member name of type `i32` in compound member access is not an instance member or an interface member.
+// CHECK:STDERR: var b: i32 = a.({.index = 1}.index);
+// CHECK:STDERR:              ^~~~~~~~~~~~~~~~~~~~~~
+var b: i32 = a.({.index = 1}.index);
```
This seems like a regression? Shouldn't a constant integer value be accepted in `a.(___)` as long as its constant value is in range?

josh11b commented on toolchain/lower/testdata/index/tuple_return_value_access.carbon:
```diff
@@ -11,34 +11,8 @@
+  // CHECK:STDERR: tuple_return_value_access.carbon:[[@LINE+3]]:16: ERROR: Type `(i32, i32)` does not support indexing.
+  // CHECK:STDERR:   var t: i32 = F()[1];
+  // CHECK:STDERR:                ^~~~~~
   var t: i32 = F()[1];
```
Why was this not changed?
```suggestion
  var t: i32 = F().1;
```

josh11b commented on toolchain/lower/testdata/let/tuple.carbon:
```diff
@@ -12,64 +12,8 @@ fn F() -> i32 {
+  // CHECK:STDERR: tuple.carbon:[[@LINE+3]]:10: ERROR: Member name of type `i32` in compound member access is not an instance member or an interface member.
+  // CHECK:STDERR:   return c.1.(1);
+  // CHECK:STDERR:          ^~~~~~~
+  return c.1.(1);
```
If the previous code worked, shouldn't this work too?

josh11b commented on toolchain/parse/handle_period.cpp:
```diff
@@ -23,13 +23,16 @@ static auto HandlePeriodOrArrow(Context& context, NodeKind node_kind,
     context.PushState(state);
     context.PushState(State::OnlyParenExpr);
     return;
+
```
```suggestion
```

josh11b commented on toolchain/check/member_access.cpp:
```diff
@@ -429,4 +428,81 @@ auto PerformCompoundMemberAccess(Context& context, Parse::NodeId node_id,
+    }
+    default: {
+      if (tuple_type_id != SemIR::TypeId::Error) {
+        CARBON_DIAGNOSTIC(
+            TupleWrongIndexSyntax, Error,
+            "Type `{0}` does not support array indexing. Did you mean tuple.0?",
+            SemIR::TypeId);
```
Should add a test that verifies that this diagnostic is being produced correctly. Also, isn't this the opposite case, where you are using tuple indexing on a non-tuple value?

brymer-meneses replied:
> Also, isn't this the opposite case, where you are using tuple indexing on a non-tuple value?

My bad, for some reason I was thinking of `tuple[0]`. Should we add a note telling the user the right way to index a tuple?

josh11b replied:
That would be great, but I wouldn't say it is required.

josh11b commented on toolchain/check/member_access.cpp:
```diff
@@ -429,4 +428,81 @@ auto PerformCompoundMemberAccess(Context& context, Parse::NodeId node_id,
+      } else if (!index_const_id.is_template()) {
+        // TODO: Decide what to do if the index is a symbolic constant.
+        CARBON_DIAGNOSTIC(TupleIndexNotConstant, Error,
+                          "Tuple index must be a constant.");
```
Is this error still produced? Looks like the test that used to trigger this case no longer does.

brymer-meneses replied:
this should be resolved by now

brymer-meneses replied:
this should be resolved by now

josh11b commented on toolchain/check/member_access.cpp:
```diff
@@ -429,4 +428,81 @@ auto PerformCompoundMemberAccess(Context& context, Parse::NodeId node_id,
+                          "Tuple index must be a constant.");
+        context.emitter().Emit(node_id, TupleIndexNotConstant);
+        index_inst_id = SemIR::InstId::BuiltinError;
+
```
```suggestion
```

josh11b commented on toolchain/check/member_access.cpp:
```diff
@@ -391,7 +391,6 @@ auto PerformMemberAccess(Context& context, Parse::NodeId node_id,
   return member_id;
 }
-
 auto PerformCompoundMemberAccess(Context& context, Parse::NodeId node_id,
```
Shouldn't this be changed to support expressions that evaluate to a constant integer? This is where I would expect things like `t.(1)` to be supported.

josh11b reviewed (commented):
Some initial feedback. I'll look again once these are addressed.

brymer-meneses committed:
Update toolchain/check/handle_name.cpp

Co-authored-by: josh11b <15258583+josh11b@users.noreply.github.com>

brymer-meneses committed:
Update toolchain/check/handle_name.cpp

Co-authored-by: josh11b <15258583+josh11b@users.noreply.github.com>

brymer-meneses committed:
Update toolchain/check/member_access.cpp

Co-authored-by: josh11b <15258583+josh11b@users.noreply.github.com>

brymer-meneses committed:
Update toolchain/parse/handle_period.cpp

Co-authored-by: josh11b <15258583+josh11b@users.noreply.github.com>

brymer-meneses committed:
Update toolchain/check/member_access.cpp

Co-authored-by: josh11b <15258583+josh11b@users.noreply.github.com>

brymer-meneses committed:
Update toolchain/lower/testdata/index/tuple_return_value_access.carbon

Co-authored-by: josh11b <15258583+josh11b@users.noreply.github.com>

brymer-meneses committed:
handle tuple indexing of paren expressions

brymer-meneses committed:
update tests

brymer-meneses commented on toolchain/check/member_access.cpp:
```diff
@@ -410,6 +409,8 @@ auto PerformCompoundMemberAccess(Context& context, Parse::NodeId node_id,
           member.type_id())) {
     member_id = PerformImplLookup(context, node_id, base_type_const_id,
                                   *assoc_type, member_id);
+  } else if (context.types().Is<SemIR::TupleType>(base_type_id)) {
```
I'm not sure if this is the right way to  disambiguate between other stuff and `tuple.(indexConstant)`.

brymer-meneses replied:
Yea, this fails, and results to a compiler crash. 

```
fn Main() {
  var non_tuple: [i32; 1] = (1,);
  var first: i32 = non_tuple.0;
}
```

EDIT: this is an unrelated issue, and i think i just fixed this https://github.com/carbon-language/carbon-lang/pull/4186/commits/133ad706654cf9d58bccee0df0145e5b73df40f5.

jonmeow replied:
I understand the desire to use `types().Is` here, but I'm going to suggest making use of the already-fetched type constant:

```suggestion
  } else if (context.constant_values().GetAsInst(base_type_const_id).Is<SemIR::TupleType>()) {
```

brymer-meneses replied:
It seems like the method `GetAsInst` does not exist in `ConstantValueStore`. I can try implementing it, I just need pointers on how.

brymer-meneses replied:
@jonmeow sorry for the ping, i think you might have overlooked this

brymer-meneses commented on toolchain/sem_ir/typed_insts.h:
```diff
@@ -967,10 +968,10 @@ struct TupleAccess {
-  static constexpr auto Kind = InstKind::TupleIndex.Define<Parse::IndexExprId>(
-      {.ir_name = "tuple_index"});
+  static constexpr auto Kind =
+      InstKind::TupleIndex.Define<Parse::NodeId>({.ir_name = "tuple_index"});
```
This is also a hack. I'm not sure if this is okay.

jonmeow replied:
This should be fine, added a suggestion for a TODO similar to what we have elsewhere (ArrayIndex is similar).

brymer-meneses committed:
add `fail_non_tuple_index.carbon`

also fix the bug the wherein the diagnostic `TupleIndexOnANonTupleType`
was not showing

brymer-meneses commented:
I think everything should be resolved by now, except the parts wherein I am confused on what to do.

brymer-meneses commented:
Also, apologies if I'm unnecessarily pinging you by re-requesting a review even if you said that you'd review this once everything gets resolved.

jonmeow commented on toolchain/check/testdata/tuples/fail_non_tuple_index.carbon:
```diff
@@ -0,0 +1,87 @@
+  // CHECK:STDERR: fail_non_tuple_index.carbon:[[@LINE+3]]:20: ERROR: Type `[i32; 1]` does not support tuple indexing. Only tuples can be indexed that way.
+  // CHECK:STDERR:   var first: i32 = non_tuple.0;
+  // CHECK:STDERR:                    ^~~~~~~~~~~
+  var first: i32 = non_tuple.0;
```
It looks like the other tuple indexing tests are under `testdata/index`. Might that be a better place for this test?

brymer-meneses replied:
I didn't notice that there was a similar test named `toolchain/check/testdata/fail_non_tuple_access`. I think I'll just merge this with that.

jonmeow replied:
Had you meant to remove this?

jonmeow commented on toolchain/lower/testdata/index/array_element_access.carbon:
```diff
@@ -13,7 +13,7 @@ fn B() -> [i32; 2] { return (1, 2); }
 fn Run() {
   var a: [i32; 2] = A();
-  var b: i32 = A()[0];
+  var b: i32 = A().0;
```
This should be focused on testing array indexing. I see this is a tuple, so how about:

```suggestion
  var b: i32 = 1;
```

This way, there's still a non-constant index for `c`.

jonmeow commented on toolchain/parse/testdata/basics/fail_invalid_designators.carbon:
```diff
@@ -15,14 +15,10 @@ fn F() {
-  // CHECK:STDERR: fail_invalid_designators.carbon:[[@LINE+3]]:5: ERROR: Expected identifier after `.`.
-  // CHECK:STDERR:   a.42;
-  // CHECK:STDERR:     ^~
   a.42;
```
This test is for invalid designators, and is expected to fail. As a consequence, if this started failing, the test would still basically pass.

What do you think of separating `a.42` out to a separate parse test, so that it's tested that this parses without an error? e.g., adding a test under parse/testdata/index for the tuple index case.

jonmeow commented on toolchain/check/testdata/index/fail_negative_indexing.carbon:
```diff
@@ -9,10 +9,10 @@
+// CHECK:STDERR: fail_negative_indexing.carbon:[[@LINE+3]]:17: ERROR: Cannot access member of interface Negate in type i32 that does not implement that interface.
+// CHECK:STDERR: var b: i32 = a.(-10);
+// CHECK:STDERR:                 ^~~
+var b: i32 = a.(-10);
```
It looks like several of these tests weren't expressly for tuples (versus ones that have array or tuple in the name, such as fail_tuple_index_error). I expect they were intended to cover indexing in general. I think that made sense when both used the same codepath, but now that tuple indexing is splitting out from array indexing more completely, should we have tests that array indexing has corresponding errors?

jonmeow commented on toolchain/parse/handle_period.cpp:
```diff
@@ -23,6 +23,8 @@ static auto HandlePeriodOrArrow(Context& context, NodeKind node_kind,
                                              NodeKind::BaseName)) {
+  } else if (context.ConsumeAndAddLeafNodeIf(Lex::TokenKind::IntLiteral,
+                                             NodeKind::IntLiteral)) {
     // OK, `.base`.
```
This comment looks like it should be associated with the "Base" case above. Maybe it'd be good to add a similar comment for IntLiteral, e.g.:

```
// OK, `.42`.
```

jonmeow commented on toolchain/parse/typed_nodes.h:
```diff
@@ -1110,6 +1110,9 @@ struct StructTypeLiteral {
 };
 
+using IntLiteral = LeafNode<NodeKind::IntLiteral, Lex::IntLiteralTokenIndex,
+                            NodeCategory::Expr | NodeCategory::IntConst>;
```
We've been trying to cluster node kinds in this file. This section is really for struct-related nodes. Can you move this to the "Literals, operators, and modifiers", around line 925, maybe putting it after the `#include` there (before `extern`)? That's where other similar literals are also defined.

jonmeow commented on toolchain/sem_ir/typed_insts.h:
```diff
@@ -967,10 +968,10 @@ struct TupleAccess {
 struct TupleIndex {
-  static constexpr auto Kind = InstKind::TupleIndex.Define<Parse::IndexExprId>(
-      {.ir_name = "tuple_index"});
+  static constexpr auto Kind =
```
```suggestion
  // TODO: Make Parse::NodeId more specific.
  static constexpr auto Kind =
```

jonmeow commented on toolchain/check/member_access.h:
```diff
@@ -23,6 +23,10 @@ auto PerformCompoundMemberAccess(Context& context, Parse::NodeId node_id,
                                  SemIR::InstId base_id,
                                  SemIR::InstId member_expr_id) -> SemIR::InstId;
 
+auto PerformTupleIndex(Context& context, Parse::NodeId node_id,
```
We try to keep comments on functions. Here's a suggestion similar to other functions in this file:

```suggestion
// Creates SemIR to perform a tuple index with base expression `tuple_inst_id` and
// index expression `index_inst_id`. Returns the result of the access.
auto PerformTupleIndex(Context& context, Parse::NodeId node_id,
```

jonmeow commented on toolchain/check/handle_name.cpp:
```diff
@@ -22,6 +23,14 @@ auto HandleParseNode(Context& context, Parse::MemberAccessExprId node_id)
     auto member_id =
         PerformCompoundMemberAccess(context, node_id, base_id, member_expr_id);
     context.node_stack().Push(node_id, member_id);
+  } else if (context.node_stack().PeekIs<Parse::NodeKind::IntLiteral>()) {
```
Here and below, I think the repeated `PeekIs` (for both ParenExpr and IntLiteral) is going to lead to the node kind being queried. To avoid the duplicate logic, can you store the valid of the kind and compare? i.e., something like:

```
auto node_kind = context.node_stack().PeekNodeKind();
if (node_kind == Parse::NodeKind::ParenExpr) {
  ...
} else if (node_kind == Parse::NodeKind::IntLiteral) {
```

jonmeow commented on toolchain/check/member_access.cpp:
```diff
@@ -429,4 +430,74 @@ auto PerformCompoundMemberAccess(Context& context, Parse::NodeId node_id,
+          context.types().TryGetAs<SemIR::TupleType>(tuple_type_id)) {
+    SemIR::TypeId element_type_id = SemIR::TypeId::Error;
+    auto index_node_id = context.insts().GetLocId(index_inst_id);
+    index_inst_id = ConvertToValueOfType(
+        context, index_node_id, index_inst_id,
+        context.GetBuiltinType(SemIR::BuiltinInstKind::IntType));
+    auto index_const_id = context.constant_values().Get(index_inst_id);
+    if (index_const_id == SemIR::ConstantId::Error) {
+      index_inst_id = SemIR::InstId::BuiltinError;
+    } else if (!index_const_id.is_template()) {
+      // TODO: Decide what to do if the index is a symbolic constant.
+      CARBON_DIAGNOSTIC(TupleIndexNotConstant, Error,
+                        "Tuple index must be a constant.");
+      context.emitter().Emit(node_id, TupleIndexNotConstant);
+      index_inst_id = SemIR::InstId::BuiltinError;
+    } else {
+      auto index_literal = context.insts().GetAs<SemIR::IntLiteral>(
+          context.constant_values().GetInstId(index_const_id));
+      auto type_block = context.type_blocks().Get(tuple_type->elements_id);
+      if (const auto* index_val = ValidateTupleIndex(
+              context, node_id, tuple_inst, index_literal, type_block.size())) {
+        element_type_id = type_block[index_val->getZExtValue()];
+      } else {
+        index_inst_id = SemIR::InstId::BuiltinError;
+      }
+    }
+
+    return context.AddInst<SemIR::TupleIndex>(node_id,
+                                              {.type_id = element_type_id,
+                                               .tuple_id = tuple_inst_id,
+                                               .index_id = index_inst_id});
+  } else {
+    CARBON_DIAGNOSTIC(TupleIndexOnANonTupleType, Error,
+                      "Type `{0}` does not support tuple indexing. Only "
+                      "tuples can be indexed that way.",
+                      SemIR::TypeId);
+    context.emitter().Emit(node_id, TupleIndexOnANonTupleType, tuple_type_id);
+  }
+
+  return SemIR::InstId::BuiltinError;
```
We tend to favor early returns to reduce nesting. Here I think you could do something like:

```suggestion
  auto tuple_type =
      context.types().TryGetAs<SemIR::TupleType>(tuple_type_id));
  if (!tuple_type) {
    CARBON_DIAGNOSTIC(TupleIndexOnANonTupleType, Error,
                      "Type `{0}` does not support tuple indexing. Only "
                      "tuples can be indexed that way.",
                      SemIR::TypeId);
    context.emitter().Emit(node_id, TupleIndexOnANonTupleType, tuple_type_id);
    return SemIR::InstId::BuiltinError;
  }
  
  SemIR::TypeId element_type_id = SemIR::TypeId::Error;
  auto index_node_id = context.insts().GetLocId(index_inst_id);
  index_inst_id = ConvertToValueOfType(
      context, index_node_id, index_inst_id,
      context.GetBuiltinType(SemIR::BuiltinInstKind::IntType));
  auto index_const_id = context.constant_values().Get(index_inst_id);
  if (index_const_id == SemIR::ConstantId::Error) {
    index_inst_id = SemIR::InstId::BuiltinError;
  } else if (!index_const_id.is_template()) {
    // TODO: Decide what to do if the index is a symbolic constant.
    CARBON_DIAGNOSTIC(TupleIndexNotConstant, Error,
                      "Tuple index must be a constant.");
    context.emitter().Emit(node_id, TupleIndexNotConstant);
    index_inst_id = SemIR::InstId::BuiltinError;
  } else {
    auto index_literal = context.insts().GetAs<SemIR::IntLiteral>(
        context.constant_values().GetInstId(index_const_id));
    auto type_block = context.type_blocks().Get(tuple_type->elements_id);
    if (const auto* index_val = ValidateTupleIndex(
            context, node_id, tuple_inst, index_literal, type_block.size())) {
      element_type_id = type_block[index_val->getZExtValue()];
    } else {
      index_inst_id = SemIR::InstId::BuiltinError;
    }
  }

  return context.AddInst<SemIR::TupleIndex>(node_id,
                                            {.type_id = element_type_id,
                                             .tuple_id = tuple_inst_id,
                                             .index_id = index_inst_id});  
```

Note, this also couple the `BuiltinError` return with the diagnostic, and I think it helps make it clearer that the other path always has a valid return.

jonmeow reviewed (commented):
Thanks, this looks really good! Josh had asked me to take a look, and I think the logic looks correct. I did see some spots though where I think it may be helpful to adjust API use, just subtle toolchain things.

jonmeow commented:
As a sidenote (not something I'm requesting for this PR), I'm thinking we may want to split up the `index` directory to reflect the logic split. Maybe leaving the `IndexWith`-y things in `index`, and moving the tuple indexing out to something like `tuples/access`. (which, is now making me notice that odd plural of "tuples": #4200)

This is partly because I feel odd seeing so many files with either `array` or `tuple` in the name, it seems like it'd be nice to have that implicit from the directory name. You can do this here if you want, but I'm happy to pick it up after you're merged since it feels like scope creep here.

brymer-meneses committed:
add comment to `PerformTupleIndex`

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

brymer-meneses committed:
add missing comments

brymer-meneses committed:
refactor toolchain/check/member_access.cpp

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

brymer-meneses committed:
move `IntLiteral`

brymer-meneses committed:
Update toolchain/check/member_access.cpp

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

brymer-meneses committed:
apply suggestion

brymer-meneses commented:
> You can do this here if you want, but I'm happy to pick it up after you're merged since it feels like scope creep here.

I'll happily do this in a separate PR, as I do not want to clutter this PR by syncing to trunk (to have https://github.com/carbon-language/carbon-lang/pull/4200).

brymer-meneses commented:
On a related note, do you think we should rename tuple index with tuple access across the repo? That would mean renaming a lot of things ( `SemIR`, `typed_nodes.h`) and stuff.

jonmeow commented:
> On a related note, do you think we should rename tuple index with tuple access across the repo? That would mean renaming a lot of things ( `SemIR`, `typed_nodes.h`) and stuff.

Asking for thoughts [on discord](https://discord.com/channels/655572317891461132/655578254970716160/1271195835975204946). Maybe best to split out any rename.

jonmeow commented:
Per discussion, we'll want to consolidate `TupleIndex` into the already-existing `TupleAccess`. They're essentially equivalent, and that'd align with how `StructAccess` works: using the same instruction for both implicit and explicit access. [ed: note this too might be better in a different PR, but I'm not sure how big the delta would be]

brymer-meneses committed:
apply suggestion

brymer-meneses committed:
consolidate `tuples/fail_non_tuple_index.carbon` with `tuples/index/fail_non_tuple_access.carbon

brymer-meneses committed:
Update toolchain/lower/testdata/index/array_element_access.carbon

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

brymer-meneses committed:
apply test changes

brymer-meneses committed:
add tuple negative indexing

brymer-meneses committed:
testdata changes

brymer-meneses committed:
testdata changes

jonmeow commented on toolchain/parse/testdata/index/tuple_index.carbon:
```diff
@@ -0,0 +1,42 @@
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/parse/testdata/tuple/index.carbon
+
+let tuple: (i32, i32) = (5, 5);
```
```suggestion
```

This is really small (I'm mainly mentioning it because of fail_non_tuple_index.carbon), but I think it's helpful to both emphasize the `tuple.0` parse tree, and as a hint that this line isn't needed (name lookup doesn't occur during parse). So the parse tree below ends up a lot shorter, making it easier to see the syntax under test.

jonmeow reviewed (approved):
Thanks for all the changes! Basically looks good, but I think you might've missed deleting the fail_non_tuple_index.carbon. Since there's fail_non_tuple_access (as you observed), can you delete that and I'll merge?

jonmeow commented:
> Thanks for all the changes! Basically looks good, but I think you might've missed deleting the fail_non_tuple_index.carbon. Since there's fail_non_tuple_access (as you observed), can you delete that and I'll merge?

Actually, it looks like things have gotten slightly out of date due to a LLVM IR change (#4212 looks like the source of conflicts). I'm happy to have you merge/autoupdate, or I can take care of that (let me know how you'd prefer).

brymer-meneses committed:
testdata changes

brymer-meneses commented:
I think, I might have messed things up when trying to sync.
